### PR TITLE
[kpm] Update garbage deletion

### PR DESCRIPTION
**Pull request by SHtress**
* [x] Read PR [documentation](https://github.com/ostis-ai/sc-machine/blob/main/CONTRIBUTING.md)
* [x] Update changelog
* [x] Update documentation

### CODE REVIEW

File: docs/sc-kpm/sc_core_agents.tex [hunk 1]
```diff
@@ -22,12 +22,15 @@
...
     \end{scnindent}
-    \scnitem{Реализация базового механизма сборки информационного мусора}
+    \scnitem{Реализация базового агента удаления множества sc-элементов}
     \begin{scnindent}
         \begin{scnrelfromlist}{используемый язык программирования}
             \scnitem{С}
         \end{scnrelfromlist}
-        \scntext{примечание}{Текущая реализация механизма сборки информационного мусора содержит один sc-агент, реагирующий на явное добавление какого-либо sc-элемента во множество \scnqqi{информационный мусор} и осуществляющий физическое удаление этого sc-элемента из sc-памяти}
+        \scniselement{атомарный sc-агент}
+        \scnrelfrom{зависимость}{Реализация sc-памяти}
```
Comment (NikitaZotov) [inline docs/sc-kpm/sc_core_agents.tex:31]:
```suggestion
        \begin{scnrelfromlist}{зависимости компонента}
                   \scnitem{Реализация sc-памяти}
        \end{scnrelfromlist}
```

File: sc-kpm/sc-utils/utils_keynodes.c [hunk 2]
```diff
@@ -4,18 +4,20 @@
...
 #include "../sc-common/sc_keynodes.h"
-#include "utils.h"
 #include <glib.h>
 
-const char keynode_sc_garbage_str[] = "sc_garbage";
+sc_addr keynode_question_erase_element;
+sc_addr keynode_init_memory_generated_structure;
 
-sc_addr keynode_sc_garbage;
+const char keynode_question_erase_element_str[] = "question_erase_element";
```
Comment (NikitaZotov) [inline sc-kpm/sc-utils/utils_keynodes.c:15]: Please, create issue to rename all questions to actions

File: sc-kpm/tests/sc-utils/sc-utils.cpp [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+TEST_F(ScMemoryTest, garbage_delete_success)
+{
+  sc_memory_context * context = sc_memory_context_new(sc_access_lvl_make_min);
+
+  search_keynodes_initialize(context);
+  sc_addr const structAddr = sc_memory_node_new(context, sc_type_node_struct | sc_type_const);
+  sc_module_initialize_with_init_memory_generated_structure(structAddr);
+
+  sc_event * event_garbage_deletion =
+      sc_event_new(context, keynode_question_initiated, SC_EVENT_ADD_OUTPUT_ARC, 0, agent_garbage_delete, 0);
```
Comment (NikitaZotov) [inline sc-kpm/tests/sc-utils/sc-utils.cpp:22]:
```suggestion
      sc_event_new(context, keynode_question_initiated, SC_EVENT_ADD_OUTPUT_ARC, null_ptr, agent_garbage_delete, null_ptr);
```

Comment (NikitaZotov) [reply sc-kpm/tests/sc-utils/sc-utils.cpp:22]: Check it everywhere

File: sc-memory/sc-core/sc_memory_ext.c [hunk 4]
```diff
@@ -23,6 +23,7 @@ typedef struct _sc_module_info
   gchar * path;
   sc_uint32 priority;
   fModuleFunc init_func;
+  char * init_func_type;
```
Comment (NikitaZotov) [inline sc-memory/sc-core/sc_memory_ext.c:26]:
```suggestion
  sc_char * init_func_type;
```

File: sc-memory/sc-core/sc_memory_ext.c [hunk 5]
```diff
@@ -127,10 +131,19 @@ sc_result sc_ext_initialize(const sc_char * ext_dir_path, const sc_char ** enabl
     // skip non module files
     if (g_str_has_suffix(file_name, moduleSuffix) == SC_TRUE)
     {
-      if (g_module_symbol(mi->ptr, "sc_module_initialize", (gpointer *)&func) == SC_FALSE)
+      if (g_module_symbol(mi->ptr, "sc_module_initialize_with_init_memory_generated_structure", (gpointer *)&func) ==
```
Comment (NikitaZotov) [inline sc-memory/sc-core/sc_memory_ext.c:134]:
```suggestion
      if (g_module_symbol(mi->ptr, "sc_module_initialize_ext", (gpointer *)&func) ==
```

File: sc-memory/sc-core/sc_memory_ext.c [hunk 6]
```diff
@@ -166,11 +179,21 @@ sc_result sc_ext_initialize(const sc_char * ext_dir_path, const sc_char ** enabl
 
   // initialize modules
   GList * item = modules_priority_list;
+  sc_result init_result = SC_RESULT_ERROR;
   while (item != null_ptr)
   {
     sc_module_info * module = (sc_module_info *)item->data;
     sc_message("Initialize module: %s", module->path);
-    if (module->init_func() != SC_RESULT_OK)
+    if (module->init_func_type == "sc_module_initialize_with_init_memory_generated_structure")
```
Comment (NikitaZotov) [inline sc-memory/sc-core/sc_memory_ext.c:187]: Compare strings in c with sc_str_cmp

**Review by NikitaZotov — CHANGES_REQUESTED** [review]

File: docs/sc-kpm/sc_core_agents.tex [hunk 7]
```diff
@@ -22,12 +22,15 @@
...
+        \scntext{пояснение}{Текущая реализация агента осуществляет физическое удаление sc-элементов, принадлежавших входному множеству, из sc-памяти, если они не принадлежат согласованной части базы знаний.}
```
[the remark this answers is not in the document]
Comment (SHtress) [reply docs/sc-kpm/sc_core_agents.tex:32]: Let's discuss format

File: sc-kpm/tests/sc-utils/sc-utils.cpp (hunk 3 above)
Comment (SHtress) [reply sc-kpm/tests/sc-utils/sc-utils.cpp:22]: Can't build after this changes

**Review by MikhailSadovsky — APPROVED** [review]

File: docs/changelog.md [hunk 8]
```diff
@@ -28,6 +28,7 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 ### Changed
 
+- Update garbage deletion agent logic, add check if element belongs to init memory structure
```
Comment (NikitaZotov) [inline docs/changelog.md:31]: Add info about fixes and info about other changes

File: docs/changelog.md [hunk 9]
```diff
@@ -37,6 +41,8 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 ### Fixed
 
+- Creating sc-keynodes without defining type. Now they will be created with const node type.
+- Converting to string params from configs([248](https://github.com/ostis-ai/sc-machine/issues/248))
```
Comment (NikitaZotov) [inline docs/changelog.md:45]:
```suggestion
- Converting to string params from configs
```

**Review by NikitaZotov — APPROVED** [review]

**Pull request merged**